### ISSUE TITLE
Document impact on performance

### ISSUE THREAD

**Issue by curiousercreative** [aiji42/next-with-split#137]
In particular, how is TTFB (time to first byte) affected when used? In general, it looks like it'll be an additional request round-trip from server to variant server. In the most common singular case of a Vercel hosted project, this should reduce down to just latency  (as all variants should have similar request processing time and middleware process should be negligible) and we can probably measure this and document it for a demo project as "adds X milliseconds to client TTFB for Vercel hosted projects".

This could be an important metric in determining implementation of split config. One could keep splits active with weighting to 0 for example if there was negligible impact.

**Comment by aiji42** [comment]
As you said, the Challenger will have an increased TTFB compared to the original because of the round trip.

I actually measured the TTFB on the [sample site](https://next-with-split.vercel.app/foo/baz) and found that the challenger increased by about 70ms.
This page is deployed on Vercel, and I am accessing it from Tokyo.

| ORIGINAL |TTFB| |
|:----|:----|:----|
|1st|52.18ms|![スクリーンショット 2021-12-14 午後9 42 07](https://user-images.githubusercontent.com/6711766/146000778-bc6ef963-ac43-4292-8a47-221aae664d25.png)|
|2nd|101.49ms|![スクリーンショット 2021-12-14 午後9 42 01](https://user-images.githubusercontent.com/6711766/146000782-6dcf2de6-3181-413b-9227-112ad5c5751b.png)|
|3rd|88.78ms|![スクリーンショット 2021-12-14 午後9 41 54](https://user-images.githubusercontent.com/6711766/146000785-342b8197-3de0-4f71-bfa0-6b6e6f44a99e.png)|
|4th|85.05ms|![スクリーンショット 2021-12-14 午後9 41 48](https://user-images.githubusercontent.com/6711766/146000788-5d615f94-d661-49de-946c-0ff40f12c185.png)|
|5th|115.07ms|![スクリーンショット 2021-12-14 午後9 41 40](https://user-images.githubusercontent.com/6711766/146000791-5345ef29-da6a-4087-97d7-62e3b5b2a974.png)|
|AVERAGE|88.51ms| |
|MEDIAN|88.65ms| |

| CHALLENGER |TTFB| |
|:----|:----|:----|
|1st|136.66ms|![スクリーンショット 2021-12-14 午後9 42 32](https://user-images.githubusercontent.com/6711766/146001515-8568f098-553c-4e11-bfac-386c25e85c91.png)|
|2nd|177.99ms|![スクリーンショット 2021-12-14 午後9 42 37](https://user-images.githubusercontent.com/6711766/146001517-ca7a8afb-0c23-4a39-b7bc-ccf487ffb37e.png)|
|3rd|136.12ms|![スクリーンショット 2021-12-14 午後9 42 44](https://user-images.githubusercontent.com/6711766/146001520-ac586146-ac7e-4bc5-9f7a-b9930a657c45.png)|
|4th|182.88ms|![スクリーンショット 2021-12-14 午後9 42 27](https://user-images.githubusercontent.com/6711766/146001523-290f23ee-c17e-4859-80e0-c9efc9560d54.png)|
|5th|165.28ms|![スクリーンショット 2021-12-14 午後9 42 19](https://user-images.githubusercontent.com/6711766/146001525-1b97c058-61af-4e7c-b08c-79c3a98a8257.png)|
|AVERAGE|159.79ms| |
|MEDIAN|162.53ms| |

There will be some differences depending on the region from which you access the site, so could you please help me with the measurement?

As a side note, the default behavior of `next-with-split` is to control that the bot is always assigned to the original. So the challenger's content will never be indexed, and the bot will always be the fastest to reach the content.

**Comment by curiousercreative** [comment]
Sure, I'll try to take some measurements. Based on your response to my other issue, my initial postulate that latency increase should be location independent needs refinement when the servers are not colocated as some requests will be client to server. As long as all variant servers are colocated, then region shouldn't impact the difference in split vs non-split, it should be fairly constant.

**Comment by curiousercreative** [comment]
I show an 84/90ms difference (median/mean) over 500 runs using apache benchmark

Original results
```
$ ab -n 500 -c 1 -C 'x-split-key-test1=original' https://next-with-split.vercel.app/foo/baz
This is ApacheBench, Version 2.3 <$Revision: 1879490 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking next-with-split.vercel.app (be patient)
Completed 100 requests
Completed 200 requests
Completed 300 requests
Completed 400 requests
Completed 500 requests
Finished 500 requests


Server Software:        Vercel
Server Hostname:        next-with-split.vercel.app
Server Port:            443
SSL/TLS Protocol:       TLSv1.2,ECDHE-RSA-AES256-GCM-SHA384,2048,256
Server Temp Key:        X25519 253 bits
TLS Server Name:        next-with-split.vercel.app

Document Path:          /foo/baz
Document Length:        23048 bytes

Concurrency Level:      1
Time taken for tests:   110.085 seconds
Complete requests:      500
Failed requests:        0
Total transferred:      11860464 bytes
HTML transferred:       11524000 bytes
Requests per second:    4.54 [#/sec] (mean)
Time per request:       220.170 [ms] (mean)
Time per request:       220.170 [ms] (mean, across all concurrent requests)
Transfer rate:          105.21 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:       58   86 174.6     70    3792
Processing:    48  134  91.4    121    1551
Waiting:       41  127  91.2    114    1545
Total:        110  220 247.1    193    5343

Percentage of the requests served within a certain time (ms)
  50%    193
  66%    203
  75%    213
  80%    220
  90%    252
  95%    311
  98%    447
  99%    793
 100%   5343 (longest request)
```

Challenger results
```
$ ab -n 500 -c 1 -C 'x-split-key-test1=challenger1' https://next-with-split.vercel.app/foo/baz
This is ApacheBench, Version 2.3 <$Revision: 1879490 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking next-with-split.vercel.app (be patient)
Completed 100 requests
Completed 200 requests
Completed 300 requests
Completed 400 requests
Completed 500 requests
Finished 500 requests


Server Software:        Vercel
Server Hostname:        next-with-split.vercel.app
Server Port:            443
SSL/TLS Protocol:       TLSv1.2,ECDHE-RSA-AES256-GCM-SHA384,2048,256
Server Temp Key:        X25519 253 bits
TLS Server Name:        next-with-split.vercel.app

Document Path:          /foo/baz
Document Length:        23778 bytes

Concurrency Level:      1
Time taken for tests:   149.415 seconds
Complete requests:      500
Failed requests:        0
Total transferred:      12243900 bytes
HTML transferred:       11889000 bytes
Requests per second:    3.35 [#/sec] (mean)
Time per request:       298.829 [ms] (mean)
Time per request:       298.829 [ms] (mean, across all concurrent requests)
Transfer rate:          80.03 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:       58   75  46.9     69    1057
Processing:   113  224 116.4    205    1908
Waiting:      105  217 116.2    197    1902
Total:        176  299 151.4    276    2965

Percentage of the requests served within a certain time (ms)
  50%    276
  66%    291
  75%    304
  80%    317
  90%    359
  95%    445
  98%    514
  99%    619
 100%   2965 (longest request)
```

**Comment by curiousercreative** [comment]
Nothing scientific, but because all of those static assets are going through a middleware hop, we're seeing that same latency increase in the browser. In my application, I think I'm seeing 2-3x the TTFB latency increase (~200ms total) for window load event. Once for the initial document, another for the client side application and then a third for any asset requests originating from the client side application.

**Comment by aiji42** [comment]
Thanks for measuring the performance. It's very detailed and interesting.

---

> Nothing scientific, but because all of those static assets are going through a middleware hop, we're seeing that same latency increase in the browser. In my application, I think I'm seeing 2-3x the TTFB latency increase (~200ms total) for window load event. Once for the initial document, another for the client side application and then a third for any asset requests originating from the client side application.

Is it possible for you to provide the URL of the project where you are experiencing many of these delays?
My sample site that I showed above is very small and is not experiencing the issues that are happening in your project. As I explained in another issue (#138), under the Next.js spec, requests for static assets do not go through the middleware, so theoretically the only delay should be in the initial HTML return, which makes your issue seem very strange.

**Comment by curiousercreative** [comment]
@aiji42 I don't have a URL to share at this time, but some logs from middleware are what led me to believe static assets are passing through there. This may be specific to our project configuration, but we have four requests (/static/fonts and /static/favicon) logged in middleware for each document requested. My previous comment was also based on anectdotal measurements in browser. I may revisit this with a browser automation measurement library run. For now, here is an update on measurements (document server response):
- 100ms native (without middleware or withSplit config)
- 160ms empty middleware, no withSplit config (just adding a middleware file adds 60ms latency)
- 160ms withSplits config + withSplit middleware and original branch selected
- 210ms withSplits config + withSplit middleware and variant branch selected

Because there's a 60ms penalty for having the middleware file in place, I've added a [prepare](https://docs.npmjs.com/cli/v8/using-npm/scripts#life-cycle-scripts) script that will install the middleware when an env flag indicating that traffic splitting is requested is set. Perhaps there's a better supported method for tapping into the build process, but this seems to work.

**Comment by aiji42** [comment]
Are `/static/fonts` and `/static/favicon` files located in the public directory? I wrote that `/_next/static` and `/_next/data` will not pass through the middleware, but files located in public directory will pass through the middleware, so be careful.

I think controlling middleware installation using the prepare command is a great idea. Could you please share the contents of that script? I would like to include it in the README.

If the target path of your A/B test is not the top page, you can minimize the number of middleware penalized requests by putting the `_middleware.ts` in the target page directory.
For example, if `/foo` is the target of your A/B test.
```
# NOT GOOD
pages
├── index.tsx
├── _middleware.ts
├── foo
│   ├── index.tsx
│   └──  ...
...

# GOOD
pages
├── index.tsx
├── foo
│   ├── index.tsx
│   ├── _middleware.ts
│   └──  ...
...
```

**Comment by curiousercreative** [comment]
I'll check these assets, thanks. I'll probably add the prepare to my demo repo and let you know.

**Comment by curiousercreative** [comment]
@aiji42 I've updated my demo repo with pretty much my full implementation for canary deploys: https://next-with-split-demo.vercel.app/. I'd be grateful for any feedback!

**Comment by aiji42** [comment]
@curiousercreative Thanks for sharing the repository. It's a very cool script!
I will generalize this script and incorporate it into this library, and I will include these insights and know-how gained by your contribution in this Issue in the README so that it will be useful to everyone who uses this library.

**Comment by curiousercreative** [comment]
Yeah, I believe almost all of it can be quite universal. My mechanism for installing middleware requires enhancement for running multiple splits with multiple custom middleware files that need to be installed to various paths. That should be simple enough to solve for, but dealing with multiple splits on the same path would be more difficult. Does this library currently support that?

**Comment by aiji42** [comment]
>  That should be simple enough to solve for, but dealing with multiple splits on the same path would be more difficult. Does this library currently support that?
Yes, I already support this, so this is not such a big issue.
My concern is that when there are other middleware in the repository other than next-with-split, I have to consider merging with it, but I will try to resolve this during the New Year vacation.

**Comment by curiousercreative** [comment]
@aiji42 yeah, I had considered the middleware merging as well and unfortunately, I came to the conclusion that this package probably shouldn't support it. I'd recommend throwing an error when you encounter existing middleware that would be overwritten and require package consumers to manage that configuration themselves. I say this because I'd feel icky rewriting version controlled source files. I can see that it would be totally possible to merge and unmerge these though, but I think it's potential full of pitfalls.

**Comment by aiji42** [comment]
@curiousercreative, I apologize for not responding for so long.
I have just released v4.3.0.
Inspired by the script you provided, I added an option to automatically install the middleware files in the original branch when the A/B test is running, and remove the middleware for the rest (challenger side or when the A/B test is stopped). It is opted out by default.
Also, the impact of this library on performance has been noted in the readme. https://github.com/aiji42/next-with-split#impact-on-performance

**Comment by curiousercreative** [comment]
@aiji42 looks great! The only comment I have is in regards to this performance documentation:
>Make sure that the middlewares do not get into routes that are not related to A/B testing.
I've found it difficult to only enable middleware for subroutes due to the UX impact of hard loads rather than prefetch and in-app navigation.
